src/release/repos.rs

```diff
@@ -1,29 +1,100 @@
 
 /// Backup the sources lists
-pub fn backup() -> anyhow::Result<()> {
-    if Path::new(BACKUP_DIR).exists() {
-        fs::remove_dir_all(BACKUP_DIR)
-            .context("failed to remove previous backup of repositories")?;
-    }
+pub fn backup(release: &str) -> anyhow::Result<()> {
+    if Path::new(PPA_DIR).exists() {
+        // Remove previous backups
+        let dir = fs::read_dir(PPA_DIR).context("cannot read PPA directory")?;
+        iter_files(dir, |entry| {
+            if entry.file_name().to_bytes().ends_with(b".save") {
+                let path = entry.path();
+                info!("removing old backup at {}", path.display());
+                fs::remove_file(&path)
+                    .with_context(|| fomat!("failed to remove backup at "(path.display())))?;
+            }
 
-    fs::create_dir_all(BACKUP_DIR).context("failed to create directory for source list backups")?;
+            Ok(())
+        })?;
 
-    if Path::new(PPA_DIR).exists() {
-        fs::rename(PPA_DIR, BACKUP_PPA_DIR).context("failed to move PPAs to backup dir")?;
+        // Create new backups
+        let dir = fs::read_dir(PPA_DIR).context("cannot read PPA directory")?;
+        iter_files(dir, |entry| {
+            let src_path = entry.path();
+            let dst_path_buf = [&*(src_path.to_bytes()), b".save"].concat();
+            let dst_path_str = OsStr::from_bytes(&dst_path_buf).unwrap();
+            let dst_path = Path::new(&dst_path_str);
+
+            info!("creating backup of {} to {}", src_path.display(), dst_path.display());
+            fs::copy(&src_path, dst_path).with_context(
+                || fomat!("failed to copy " (src_path.display()) " to " (dst_path.display())),
+            )?;
+
+            Ok(())
+        })?;
     }
 
-    fs::rename(MAIN_FILE, BACKUP_MAIN_FILE).context("failed to move sources list to backup dir")?;
+    if Path::new(MAIN_FILE).exists() {
+        if Path::new(BACKUP_MAIN_FILE).exists() {
+            info!("removing old backup at {}", BACKUP_MAIN_FILE);
+            fs::remove_file(BACKUP_MAIN_FILE).context("failed to remove backup of sources.list")?;
+        }
+        info!("creating backup of {} to {}", MAIN_FILE, BACKUP_MAIN_FILE);
+        fs::copy(MAIN_FILE, BACKUP_MAIN_FILE)
+            .context("failed to copy sources list to backup path")
+            .map(|_| ())
+    } else {
+        info!("sources list was not found — creating a new one");
+        sources::create_new_sources_list(release).context("failed to create new sources.list")
+    }
+}
 
-    Ok(())
+/// For each `.list` in `sources.list.d`, add `#` to the `deb` lines.
+pub fn disable_third_parties() -> anyhow::Result<()> {
+    let dir = fs::read_dir(PPA_DIR).context("cannot read PPA directory")?;
+    iter_files(dir, |entry| {
+        if entry.file_name().to_bytes().ends_with(b".list") {
+            let path = entry.path();
+
+            if let Some(fname) = path.file_name() {
```

Comment:
I had just looked at my `sources.list.d/system76-ubuntu-pop-focal.list` file, so I know it was present. After attempting to run an upgrade, I got this output:

```
system76@pop-os:~$ pop-upgrade release upgrade -f
Current Release: 20.04
Upgrading to: 20.10
New version available: false
Release upgrade status: release upgrade aborted: unable to disable third party repositories: failed to read /etc/apt/sources.list.d/system76-ubuntu-pop-focal.list
pop-upgrade: dbus client error: calling ReleaseUpgradeFinalize method failed
```

The `sources.list.d/system76-ubuntu-pop-focal.list` file is gone, and trying to upgrade again gives the same output.

- We probably shouldn't fail if it's not found. We've been inconsistent with where we're placing the PPA, and there could be valid configurations where this file isn't present to remove.
- This section to remove the system76-ubuntu-pop file should probably be below the section where sources are disabled (I would assume it's easier to disable everything in sources.list.d first and then remove this particular file, rather than trying to remove this one first and then disable everything except for this one, but whichever way works should be fine.)

Reply:
Oh I see. It's removed and then attempted to back up.